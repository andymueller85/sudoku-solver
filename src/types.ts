export type Row = Array<string>
export type Grid = Array<Row>

export type PossiblesRow = Array<Array<string>>
export type PossiblesGrid = Array<PossiblesRow>

export type GridWithMeta = { grid: Grid | undefined; recursiveIterations: number }
export type Matches = { possibles: Array<string>; matches: Array<number> }
