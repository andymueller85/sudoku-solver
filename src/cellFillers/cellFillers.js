import lodash from 'lodash'
import {
  columnNeighborsAreFilled,
  columnNeighborsContainNumber,
  getBoxCurColumn,
  getBoxCurRow,
  getBoxMissingCells,
  getBoxMissingNums,
  getBoxTopLeft,
  getColumnFilledNums,
  getColumnMissingNums,
  getMissingNums,
  getRowMissingCells,
  getRowMissingNums,
  isFilled,
  rowNeighborsAreFilled,
  rowNeighborsContainNumber,
  swapXY
} from '../helpers/helpers.js'

const { cloneDeep } = lodash

export function applyDefinites(grid, possibleValsGrid) {
  let myGrid = cloneDeep(grid)

  possibleValsGrid.forEach((r, rIdx) =>
    r.forEach((c, cIdx) => {
      if (!isFilled(myGrid[rIdx][cIdx])) {
        if (c.length === 1) myGrid[rIdx][cIdx] = c[0]
      }
    })
  )

  return myGrid
}

/************** cell-filling fns - rows & cols ****************/
export function cellCanBeDeterminedForRow(grid, rowNum, colNum, num) {
  const rowMissingNums = getMissingNums(grid[rowNum])
  const columnMissingNums = getColumnMissingNums(grid, colNum)

  return (
    (rowMissingNums.length === 1 && rowMissingNums[0] === num) ||
    (rowMissingNums.includes(num) &&
      columnMissingNums.includes(num) &&
      getRowMissingCells(grid, rowNum)
        .filter(cell => cell !== colNum)
        .every(v => getColumnFilledNums(grid, v).includes(num)) &&
      getBoxMissingNums(grid, getBoxTopLeft(rowNum), getBoxTopLeft(colNum)).includes(num))
  )
}

export function fillRows(grid) {
  let myGrid = cloneDeep(grid)

  myGrid.forEach((_, rowNum) => {
    getRowMissingNums(myGrid, rowNum).forEach(n => {
      getRowMissingCells(myGrid, rowNum).forEach(colNum => {
        if (cellCanBeDeterminedForRow(myGrid, rowNum, colNum, n)) {
          myGrid[rowNum][colNum] = n
        }
      })
    })
  })

  return myGrid
}

// aka Phil Collims
export function fillColumns(grid) {
  return swapXY(fillRows(swapXY(grid)))
}

/************** cell-filling fns - boxes ****************/
export function cellCanBeDeterminedForBox(grid, cell, topLeftRow, topLeftCol, num) {
  const curRow = getBoxCurRow(topLeftRow, cell)
  const curCol = getBoxCurColumn(topLeftCol, cell)
  const rNeighborsFilled = rowNeighborsAreFilled(grid, topLeftCol, curRow, curCol)
  const rNeighborsContainNum = rowNeighborsContainNumber(grid, topLeftRow, curRow, num)
  const cNeighborsAreFilled = columnNeighborsAreFilled(grid, topLeftRow, curRow, curCol)
  const cNeighborsContainNum = columnNeighborsContainNumber(grid, topLeftCol, curCol, num)
  const boxMissingNums = getBoxMissingNums(grid, topLeftRow, topLeftCol)

  return (
    // this is the only missing number - we're done.
    (boxMissingNums.length === 1 && boxMissingNums[0] === num) ||
    // bail if number is already in the box, row, or column
    (boxMissingNums.includes(num) &&
      getRowMissingNums(grid, curRow).includes(num) &&
      getColumnMissingNums(grid, curCol).includes(num) &&
      // Both other column cells in box are determined and both other complete columns have the number
      ((cNeighborsAreFilled && cNeighborsContainNum) ||
        // Both other row cells in box are determined and both other complete rows have the number
        (rNeighborsFilled && rNeighborsContainNum) ||
        // Both other complete rows and both other complete columns have the number
        (rNeighborsContainNum && cNeighborsContainNum)))
  )
}

export function fillBoxes(grid) {
  // TODO: it would probably be cleaner to flatten the boxes
  let myGrid = cloneDeep(grid)

  for (let topLeftRow = 0; topLeftRow < myGrid.length; topLeftRow += 3) {
    for (let topLeftCol = 0; topLeftCol < myGrid[0].length; topLeftCol += 3) {
      getBoxMissingNums(myGrid, topLeftRow, topLeftCol).forEach(n => {
        getBoxMissingCells(myGrid, topLeftRow, topLeftCol).forEach(cell => {
          const curRow = getBoxCurRow(topLeftRow, cell)
          const curColumn = getBoxCurColumn(topLeftCol, cell)

          if (cellCanBeDeterminedForBox(myGrid, cell, topLeftRow, topLeftCol, n))
            myGrid[curRow][curColumn] = n
        })
      })
    }
  }

  return myGrid
}
