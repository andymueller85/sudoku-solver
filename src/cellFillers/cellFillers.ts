import lodash from 'lodash'
import {
  columnNeighborsAreFilled,
  columnNeighborsContainNumber,
  getBoxCurColumn,
  getBoxCurRow,
  getBoxMissingCells,
  getBoxMissingNums,
  getBoxTopLeft,
  getColumnFilledNums,
  getColumnMissingNums,
  getMissingNums,
  getRowMissingCells,
  getRowMissingNums,
  rowNeighborsAreFilled,
  rowNeighborsContainNumber,
  swapXY
} from '../helpers/helpers'
import { Grid, SudokuNumber } from '../types'

const { cloneDeep } = lodash

export function cellCanBeDeterminedForRow(
  grid: Grid,
  rowNum: number,
  colNum: number,
  num: SudokuNumber
) {
  const rowMissingNums = getMissingNums(grid[rowNum])
  const columnMissingNums = getColumnMissingNums(grid, colNum)

  return (
    (rowMissingNums.length === 1 && rowMissingNums[0] === num) ||
    (rowMissingNums.includes(num) &&
      columnMissingNums.includes(num) &&
      getRowMissingCells(grid, rowNum)
        .filter(cell => cell !== colNum)
        .every(v => getColumnFilledNums(grid, v).includes(num)) &&
      getBoxMissingNums(grid, getBoxTopLeft(rowNum), getBoxTopLeft(colNum)).includes(num))
  )
}

export function fillRows(grid: Grid) {
  let myGrid = cloneDeep(grid)

  myGrid.forEach((_, rowNum) => {
    getRowMissingNums(myGrid, rowNum).forEach(n => {
      getRowMissingCells(myGrid, rowNum).forEach(colNum => {
        if (cellCanBeDeterminedForRow(myGrid, rowNum, colNum, n)) {
          myGrid[rowNum][colNum] = n
        }
      })
    })
  })

  return myGrid
}

/* aka Phil Collims */
export function fillColumns(grid: Grid) {
  return swapXY(fillRows(swapXY(grid)))
}

export function cellCanBeDeterminedForBox(
  grid: Grid,
  cell: number,
  topLeftRow: number,
  topLeftCol: number,
  num: SudokuNumber
) {
  const curRow = getBoxCurRow(topLeftRow, cell)
  const curCol = getBoxCurColumn(topLeftCol, cell)
  const rNeighborsFilled = rowNeighborsAreFilled(grid, curRow, curCol)
  const rNeighborsContainNum = rowNeighborsContainNumber(grid, curRow, num)
  const cNeighborsAreFilled = columnNeighborsAreFilled(grid, curRow, curCol)
  const cNeighborsContainNum = columnNeighborsContainNumber(grid, curCol, num)
  const boxMissingNums = getBoxMissingNums(grid, topLeftRow, topLeftCol)

  return (
    // this is the only missing number - we're done.
    (boxMissingNums.length === 1 && boxMissingNums[0] === num) ||
    // bail if number is already in the box, row, or column
    (boxMissingNums.includes(num) &&
      getRowMissingNums(grid, curRow).includes(num) &&
      getColumnMissingNums(grid, curCol).includes(num) &&
      // Both other column cells in box are determined and both other complete columns have the number
      ((cNeighborsAreFilled && cNeighborsContainNum) ||
        // Both other row cells in box are determined and both other complete rows have the number
        (rNeighborsFilled && rNeighborsContainNum) ||
        // Both other complete rows and both other complete columns have the number
        (rNeighborsContainNum && cNeighborsContainNum)))
  )
}

export function fillBoxes(grid: Grid) {
  // TODO: it would probably be cleaner to flatten the boxes
  let myGrid = cloneDeep(grid)

  for (let topLeftRow = 0; topLeftRow < myGrid.length; topLeftRow += 3) {
    for (let topLeftCol = 0; topLeftCol < myGrid[0].length; topLeftCol += 3) {
      getBoxMissingNums(myGrid, topLeftRow, topLeftCol).forEach(n => {
        getBoxMissingCells(myGrid, topLeftRow, topLeftCol).forEach(cell => {
          const curRow = getBoxCurRow(topLeftRow, cell)
          const curColumn = getBoxCurColumn(topLeftCol, cell)

          if (cellCanBeDeterminedForBox(myGrid, cell, topLeftRow, topLeftCol, n))
            myGrid[curRow][curColumn] = n
        })
      })
    }
  }

  return myGrid
}
