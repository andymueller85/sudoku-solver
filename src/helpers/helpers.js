export const GRID_SIZE = 9
export const possibleNums = '123456789'.split('')
export const allIndexes = Array.from({ length: GRID_SIZE }, (_, i) => i)
export const boxIndexes = [0, 3, 6]
export const miniGridIndexes = [0, 1, 2]

//************** General helper fns ****************/
export function isFilled(cell) {
  return cell !== '.'
}

export function getFilledCellCount(grid) {
  return grid.map(r => r.filter(c => isFilled(c))).flat().length
}

export function arrayIntersection(...arrays) {
  return arrays.reduce((acc, cur) => acc.filter(a => cur.includes(a)))
}

export function getColumnArray(grid, colNum) {
  return grid.map(r => r[colNum])
}

export function swapXY(grid) {
  return allIndexes.map(c => grid.map(r => r[c]))
}

export function allArraysAreEqual(...arrayOfArrays) {
  const firstArr = arrayOfArrays[0]

  return arrayOfArrays
    .slice(1)
    .every(
      a => a.length === firstArr.length && a.every((element, idx) => element === firstArr[idx])
    )
}

export function getUniqueArrays(...arrayOfArrays) {
  let uniqueArrays = [arrayOfArrays[0]]

  arrayOfArrays.slice(1).forEach(a => {
    if (uniqueArrays.every(u => !allArraysAreEqual(u, a))) {
      uniqueArrays.push(a)
    }
  })

  return uniqueArrays
}

export function gridHasAnyDeadEnds(grid) {
  return grid.some((r, rIdx) =>
    r.some(
      (_, cIdx) =>
        !isFilled(grid[rIdx][cIdx]) && getPossibleCellValues(grid, rIdx, cIdx).length === 0
    )
  )
}

export function getNextEmptyCellCoordinates(grid, rowNum, colNum) {
  let newRowNum = rowNum
  let newColNum = colNum
  const lastIndex = GRID_SIZE - 1

  if (rowNum === lastIndex && colNum === lastIndex) return undefined

  if (colNum === lastIndex) {
    newRowNum += 1
    newColNum = 0
  } else {
    newColNum += 1
  }

  if (!isFilled(grid[newRowNum][newColNum])) return [newRowNum, newColNum]

  return getNextEmptyCellCoordinates(grid, newRowNum, newColNum)
}

export function stringifyGrid(grid, isPossibleVals = false) {
  const vertSeparator = '|'
  const innerSeparator = `  ${vertSeparator}  `
  const rowBegin = `${vertSeparator}  `
  const rowEnd = `  ${vertSeparator}`
  const pad = rowBegin.length + rowEnd.length
  const horSeparator = '-'
  const boxSeparatorVert = String.fromCharCode(parseInt('275A', 16))
  const boxSeparatorHor = isPossibleVals ? horSeparator : '='

  function replaceCharAt(str, i, char) {
    return str.substring(0, i) + char + str.substring(i + 1)
  }

  let strGrid = horSeparator.repeat((GRID_SIZE - 1) * innerSeparator.length + GRID_SIZE + pad)

  for (let row = 0; row < GRID_SIZE; row++) {
    let logStr = grid[row].map(c => (isFilled(c) ? c : ' ')).join(innerSeparator)

    if (!isPossibleVals) {
      logStr = replaceCharAt(logStr, 15, boxSeparatorVert)
      logStr = replaceCharAt(logStr, 33, boxSeparatorVert)
    }

    strGrid += `\n${rowBegin}${logStr}${rowEnd}\n`
    strGrid += ([2, 5].includes(row) ? boxSeparatorHor : horSeparator).repeat(logStr.length + pad)
  }

  return strGrid
}

/* istanbul ignore next */
export function printGrid(grid) {
  console.log(stringifyGrid(grid))
}

export function seedGrid(input) {
  return input
    .split(/\r?\n/)
    .filter(r => r)
    .map(r => [...r].map(c => c))
}

export function getGridPossibleValues(grid) {
  return grid.map((r, rIdx) =>
    r.map((c, cIdx) => {
      if (isFilled(c)) return [c]

      const rowMissingNums = getRowMissingNums(grid, rIdx)
      const colMissingNums = getColumnMissingNums(grid, cIdx)
      const boxMissingNums = getBoxMissingNums(grid, getBoxTopLeft(rIdx), getBoxTopLeft(cIdx))

      return arrayIntersection(rowMissingNums, colMissingNums, boxMissingNums)
    })
  )
}

//************** Box helper fns ****************/
export function getBoxTopLeft(rowOrCol) {
  return rowOrCol - (rowOrCol % 3)
}

export function getBoxCurRow(topLeftRow, cell) {
  return topLeftRow + Math.floor(cell / 3)
}

export function getBoxCurColumn(topLeftCol, cell) {
  return topLeftCol + (cell % 3)
}

export function getTopLeftColumnForBoxNum(boxNum) {
  return 3 * (boxNum % 3)
}

export function getBoxTopLeftCoordinates(boxNum) {
  return [getBoxTopLeft(boxNum), getTopLeftColumnForBoxNum(boxNum)]
}

export function getBoxIndexes(topLeftRowOrColumn) {
  return [topLeftRowOrColumn, topLeftRowOrColumn + 1, topLeftRowOrColumn + 2]
}

export function flattenBox(grid, topLeftRowNum, topLeftColNum) {
  return grid
    .slice(topLeftRowNum, topLeftRowNum + 3)
    .map(r => r.slice(topLeftColNum, topLeftColNum + 3))
    .flat()
}

export function unflattenBox(boxArray) {
  return boxIndexes.map(i => boxArray.slice(i, i + 3))
}

export function flattenBoxes(grid) {
  return boxIndexes.map(r => boxIndexes.map(c => flattenBox(grid, r, c))).flat()
}

export function unflattenBoxes(grid) {
  let unflattenedGrid = Array.from({ length: GRID_SIZE }, () => Array.from({ length: GRID_SIZE }))

  grid.forEach((b, bIdx) => {
    const [topLeftRow, topLeftCol] = getBoxTopLeftCoordinates(bIdx)

    unflattenBox(b).forEach((r, rIdx) => {
      r.forEach((c, cIdx) => {
        unflattenedGrid[topLeftRow + rIdx][topLeftCol + cIdx] = c
      })
    })
  })

  return unflattenedGrid
}

export function rowNeighborsAreFilled(grid, curRow, curCol) {
  const topLeftCol = getBoxTopLeft(curCol)

  return grid[curRow]
    .map((cell, index) => ({ cell, index }))
    .slice(topLeftCol, topLeftCol + 3)
    .filter(({ index }) => index !== curCol)
    .every(({ cell }) => isFilled(cell))
}

export function columnNeighborsAreFilled(grid, curRow, curCol) {
  return rowNeighborsAreFilled(swapXY(grid), curCol, curRow)
}

export function rowNeighborsContainNumber(grid, curRow, num) {
  return getBoxIndexes(getBoxTopLeft(curRow))
    .filter(r => r !== curRow)
    .every(r => getFilledNums(grid[r]).includes(num))
}

export function columnNeighborsContainNumber(grid, curCol, num) {
  return getBoxIndexes(getBoxTopLeft(curCol))
    .filter(c => c !== curCol)
    .every(c => getColumnFilledNums(grid, c).includes(num))
}

export function getPossibleCellValues(grid, rowNum, colNum) {
  const rowMissingNums = getRowMissingNums(grid, rowNum)
  const colMissingNums = getColumnMissingNums(grid, colNum)
  const boxMissingNums = getBoxMissingNums(grid, getBoxTopLeft(rowNum), getBoxTopLeft(colNum))

  return arrayIntersection(rowMissingNums, colMissingNums, boxMissingNums)
}

//************** Filled Nums fns ****************/
export function getFilledNums(arr) {
  return arr.filter(c => isFilled(c))
}

export function getRowFilledNums(grid, rowNum) {
  return getFilledNums(grid[rowNum])
}

export function getColumnFilledNums(grid, colNum) {
  return getFilledNums(getColumnArray(grid, colNum))
}

export function getBoxFilledNums(grid, topLeftRowNum, topLeftColNum) {
  return getFilledNums(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

//************** Missing Nums fns ****************/
export function getMissingNums(arr) {
  return possibleNums.filter(n => !getFilledNums(arr).includes(n))
}

export function getRowMissingNums(grid, rowNum) {
  return getMissingNums(grid[rowNum])
}

export function getColumnMissingNums(grid, colNum) {
  return getMissingNums(getColumnArray(grid, colNum))
}

export function getBoxMissingNums(grid, topLeftRowNum, topLeftColNum) {
  return getMissingNums(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

//************** Missing Cells fns ****************/
export function getMissingCells(arr) {
  return arr
    .map((cell, index) => ({ cell, index }))
    .filter(({ cell }) => !isFilled(cell))
    .map(({ index }) => index)
}

export function getRowMissingCells(grid, rowNum) {
  return getMissingCells(grid[rowNum])
}

export function getColumnMissingCells(grid, colNum) {
  return getMissingCells(getColumnArray(grid, colNum))
}

export function getBoxMissingCells(grid, topLeftRowNum, topLeftColNum) {
  return getMissingCells(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

//************** validity fns ****************/
export function isValid(arr) {
  const filledNums = getFilledNums(arr)

  return (
    allArraysAreEqual(filledNums, [...new Set(filledNums)]) &&
    filledNums.every(n => possibleNums.includes(n))
  )
}

export function rowIsValid(grid, rowNum) {
  return isValid(grid[rowNum])
}

export function everyRowIsValid(grid) {
  return grid.every((_, rowNum) => rowIsValid(grid, rowNum))
}

export function columnIsValid(grid, colNum) {
  return isValid(getColumnArray(grid, colNum))
}

export function everyColumnIsValid(grid) {
  return everyRowIsValid(swapXY(grid))
}

export function boxIsValid(grid, topLeftRowNum, topLeftColNum) {
  return isValid(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

export function everyBoxIsValid(grid) {
  return everyRowIsValid(flattenBoxes(grid))
}

export function gridIsValid(grid) {
  return everyRowIsValid(grid) && everyColumnIsValid(grid) && everyBoxIsValid(grid)
}

//************** completeness fns ****************/
export function isComplete(arr) {
  return isValid(arr) && arr.length === GRID_SIZE && arr.every(c => isFilled(c))
}

export function rowIsComplete(grid, rowNum) {
  return isComplete(grid[rowNum])
}

export function columnIsComplete(grid, colNum) {
  return isComplete(getColumnArray(grid, colNum))
}

export function boxIsComplete(grid, topLeftRowNum, topLeftColNum) {
  return isComplete(flattenBox(grid, topLeftRowNum, topLeftColNum))
}

export function gridIsComplete(grid) {
  return gridIsValid(grid) && grid.every(r => r.every(c => isFilled(c)))
}
