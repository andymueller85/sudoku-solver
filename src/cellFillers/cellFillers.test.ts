import fs from 'fs'
import lodash from 'lodash'
import { seedGrid, stringifyGrid } from '../helpers/helpers'
import { EMPTY_CELL, SudokuNumber } from '../types'
import * as cellFillers from './cellFillers'

const { cloneDeep } = lodash
const fileInput = fs.readFileSync('./input.txt', 'utf8')
const grid = seedGrid(fileInput)

const firstBoxCompleteGrid = grid.map((r, rowI) =>
  r.map((c, colI) => (rowI < 3 && colI < 3 ? (`${rowI * 3 + colI + 1}` as SudokuNumber) : c))
)

describe('Cell-filling functions - rows and columns', () => {
  describe('cellCanBeDeterminedForRow', () => {
    const { cellCanBeDeterminedForRow } = cellFillers
    const gridWithFirstRowOneCellLeft = grid.map((r, rowIndex) =>
      r.map((c, cIndex) =>
        rowIndex === 0 ? (cIndex > 0 ? (`${cIndex + 1}` as SudokuNumber) : EMPTY_CELL) : c
      )
    )

    test('should return false if the box already contains the number', () => {
      expect(cellCanBeDeterminedForRow(grid, 1, 1, SudokuNumber.SEVEN)).toBe(false)
    })

    test('should return false if the number is already in the current row', () => {
      expect(cellCanBeDeterminedForRow(grid, 1, 2, SudokuNumber.SIX)).toBe(false)
    })

    test('should return false if there is a column that does not contain the number ', () => {
      expect(cellCanBeDeterminedForRow(grid, 2, 0, SudokuNumber.FIVE)).toBe(false)
    })

    test('should return false if the box already includes the number', () => {
      expect(cellCanBeDeterminedForRow(grid, 1, 4, SudokuNumber.FOUR)).toBe(false)
    })

    test('should return true if every other empty cell column contains the number - X axis', () => {
      expect(cellCanBeDeterminedForRow(grid, 4, 0, SudokuNumber.SEVEN)).toBe(true)
    })

    test('should return true if number is only one left for the row', () => {
      expect(cellCanBeDeterminedForRow(gridWithFirstRowOneCellLeft, 0, 0, SudokuNumber.ONE)).toBe(
        true
      )
    })

    describe('fillColumns (phil collims?)', () => {
      const { fillColumns } = cellFillers

      test('should match snapshot after one pass', () => {
        expect(stringifyGrid(fillColumns(grid))).toMatchSnapshot()
      })
    })
  })

  describe('fillRows', () => {
    const { fillRows } = cellFillers

    test('should match snapshot after one pass', () => {
      expect(stringifyGrid(fillRows(grid))).toMatchSnapshot()
    })
  })
})

describe('Cell-filling functions - boxes', () => {
  describe('cellCanBeDeterminedForBox', () => {
    const { cellCanBeDeterminedForBox } = cellFillers
    const firstBoxOneCellEmptyGrid = cloneDeep(firstBoxCompleteGrid)
    firstBoxOneCellEmptyGrid[0][0] = EMPTY_CELL
    const firstBoxFirstTwoCellsEmptyGrid = cloneDeep(firstBoxCompleteGrid)
    firstBoxFirstTwoCellsEmptyGrid[0][0] = EMPTY_CELL
    firstBoxFirstTwoCellsEmptyGrid[0][1] = EMPTY_CELL

    // cases are mostly tested by checkCellAGainstOtherAxis tests, just test a couple cases here.
    test('should return true if number is only one left for the box', () => {
      expect(cellCanBeDeterminedForBox(firstBoxOneCellEmptyGrid, 0, 0, 0, SudokuNumber.ONE)).toBe(
        true
      )
    })

    test('should return false if not', () => {
      expect(
        cellCanBeDeterminedForBox(firstBoxFirstTwoCellsEmptyGrid, 0, 0, 0, SudokuNumber.ONE)
      ).toBe(false)
    })
  })

  describe('fillBoxes', () => {
    const { fillBoxes } = cellFillers

    test('should match snapshot after one pass', () => {
      expect(stringifyGrid(fillBoxes(grid))).toMatchSnapshot()
    })
  })
})
