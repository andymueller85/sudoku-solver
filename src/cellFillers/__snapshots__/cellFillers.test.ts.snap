// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell-filling functions - boxes fillBoxes should match snapshot after one pass 1`] = `
"-------------------------------------------------------
|  9  |  1  |  6  ❚  4  |     |     ❚     |  3  |  7  |
-------------------------------------------------------
|  2  |     |     ❚     |     |     ❚  6  |     |     |
-------------------------------------------------------
|     |  8  |  7  ❚     |  3  |  6  ❚  1  |     |     |
=======================================================
|     |  9  |  5  ❚     |     |     ❚     |     |     |
-------------------------------------------------------
|     |  4  |  1  ❚  9  |  8  |  5  ❚  3  |  6  |     |
-------------------------------------------------------
|     |     |     ❚     |     |     ❚  4  |  9  |  5  |
=======================================================
|     |     |  3  ❚  1  |  9  |     ❚  8  |  2  |     |
-------------------------------------------------------
|  1  |     |  9  ❚     |     |     ❚     |     |  3  |
-------------------------------------------------------
|  4  |  7  |     ❚     |     |  3  ❚  9  |  1  |  6  |
-------------------------------------------------------"
`;

exports[`Cell-filling functions - rows and columns cellCanBeDeterminedForRow fillColumns (phil collims?) should match snapshot after one pass 1`] = `
"-------------------------------------------------------
|  9  |  1  |     ❚  4  |     |     ❚     |  3  |  7  |
-------------------------------------------------------
|  2  |     |  4  ❚     |     |  9  ❚  6  |     |     |
-------------------------------------------------------
|     |  8  |  7  ❚     |  3  |  6  ❚  1  |     |  9  |
=======================================================
|     |  9  |  5  ❚     |     |     ❚     |     |     |
-------------------------------------------------------
|     |  4  |  1  ❚  9  |  8  |  5  ❚  3  |  6  |     |
-------------------------------------------------------
|     |     |     ❚     |     |     ❚  4  |  9  |     |
=======================================================
|     |     |  3  ❚  1  |  9  |     ❚  8  |  2  |     |
-------------------------------------------------------
|     |     |  9  ❚     |     |     ❚     |     |  3  |
-------------------------------------------------------
|  4  |  7  |     ❚     |     |  3  ❚  9  |  1  |  6  |
-------------------------------------------------------"
`;

exports[`Cell-filling functions - rows and columns fillRows should match snapshot after one pass 1`] = `
"-------------------------------------------------------
|  9  |  1  |     ❚  4  |     |     ❚     |  3  |  7  |
-------------------------------------------------------
|  2  |     |     ❚     |     |     ❚  6  |     |     |
-------------------------------------------------------
|     |  8  |  7  ❚     |  3  |  6  ❚  1  |     |  9  |
=======================================================
|     |  9  |  5  ❚     |     |     ❚     |     |     |
-------------------------------------------------------
|  7  |  4  |  1  ❚  9  |  8  |  5  ❚  3  |  6  |  2  |
-------------------------------------------------------
|     |     |     ❚     |     |     ❚  4  |  9  |     |
=======================================================
|     |     |  3  ❚  1  |  9  |  7  ❚  8  |  2  |     |
-------------------------------------------------------
|     |     |  9  ❚     |     |     ❚     |     |  3  |
-------------------------------------------------------
|  4  |  7  |     ❚     |     |  3  ❚  9  |  1  |  6  |
-------------------------------------------------------"
`;
